### PR TITLE
Fix issue #72 - Check for overlapping child nodes on deletion

### PR DESCRIPTION
**Pull request by iwismer**
This pull request addresses #72 Which causes a key error due to improper balancing on node removal.

### CODE REVIEW

**Comment by iwismer** [comment]
This fixes the test case found in #72 but I have found another case where it still fails. I will investigate and try to provide a solution for that case as well.

File: intervaltree/node.py [hunk 1]
```diff
@@ -375,6 +375,13 @@ def get_new_s_center():
             child = Node(new_x_center, get_new_s_center())
             self.s_center -= child.s_center
 
+            # Move any overlaps into child
+            for iv in set(self.s_center
+                    | (self[0].s_center if self[0] else set())):
+                if iv.contains_point(child.x_center):
+                    self.s_center.discard(iv)
```
Comment (chaimleib) [inline intervaltree/node.py]: This will end up not removing `iv` from `self[0]`, and will break our invariants.

File: intervaltree/node.py [hunk 2]
```diff
@@ -397,9 +404,14 @@ def get_new_s_center():
...
-            for iv in set(new_self.s_center):
+            # This include overlapping child intervals or overlap intervals
+            for iv in set(new_self.s_center
+                    | (new_self[0].s_center if new_self[0] else set())
+                    | (new_self[1].s_center if new_self[1] else set())):
                 if iv.contains_point(greatest_child.x_center):
-                    new_self.s_center.remove(iv)
+                    # use discard instead of remove so no error is thrown
+                    # if it' not an from the center
+                    new_self.s_center.discard(iv)
```
Comment (chaimleib) [inline intervaltree/node.py]: This also forgets to remove from children of `new_self`

**Comment by alexreg** [comment]
Hi @chaimleib. Do you plan to make the necessary fixes yourself, or wait for the PR author to do them? I'm running into the bug this attempts to fix, I believe, and want to get if solved as soon as possible... unfortunately I don't quite understand what is needed.

**Comment by iwismer** [comment]
I can hopefully make the changes next week, but @chaimleib if you know how to fix those issues you can go ahead and do it.

**Comment by alexreg** [comment]
Any progress on this?

**Comment by iwismer** [comment]
I've spent a little time working on this, but it looks like it'll be a little more complex than anticipated. Removing the nodes from their old nodes and into the new s_center can't be don't normally because their location is technically incorrect (which is why they need to be moved) so they have to be removed manually and then the tree rebalanced. I'm not sure when I'll next get a chance to put some time into this. Hopefully sooner rather than later.

**Comment by alexreg** [comment]
Any progress? This is kind of a show-stopper for using this library.

**Comment by chaimleib** [comment]
The issue was fixed with the release of v3!

**Comment by alexreg** [comment]
@chaimleib Glad to hear, thanks!

**Pull request closed without merging**